### DISCUSSION TITLE
Datasync client PullAsync not supporting multiple or operations.

### DISCUSSION THREAD

**Discussion by jlantech** [Azure/azure-mobile-apps#635] [Q&A] [answered]
Hello all.

 I'm porting our Enterprise application from 4.2.0 to the new Datasync version.
We use in the old Mobile server dual call to filter specific tables for example passing the USER ID in the pull operation to get the customers from the main server and in the API server we query the first table that is a filter called CUSTOMER_FILTER where we get the customers assigned to the current DIRVER ID and them query the CUSTOMER table with the list of customers obtained from CUSOMER_FILER.

Now that is not possible on the new Datasync API. But this new approach use the oDATA filter.
I try to mimic this using oDATA filter like CUSTOMER_ID in ['121','343','342','23423'] but I get is a syntax error.
I cannot use the IN operator on the client side of the DataSync?

 Thanks for any help.

**Comment by adrianhall** [comment, marked as answer]
Just to be clear, are you trying to synchronize a table but limit the synchronized data to the list of customer IDs?

**Reply by jlantech** [reply]
Yes. I was, until I found your solution for this at the last page of the forum using the IRepository interface. Thank you.
The current issue is the requirement to pass the userId like a regular parameter but the PulIItemSync() do not have the option for that, unless I looking at the wrong place.  Trying to add the userId on the Middleware.

**Reply by adrianhall** [reply]
You can do this relatively easily.  I'd recommend actually using authentication though.  Firstly, set up an access control provider, which can use the IHttpContextAccessor - see https://learn.microsoft.com/en-us/azure/developer/mobile-apps/azure-mobile-apps/howto/server/dotnet-core#configure-access-permissions for an example.  Now, importantly, the HttpContext has access to HttpRequest, which has access to all the parameters in the request.

In your Datasync Client, you can use PullItemsAsync with a query.  Something like:

``` csharp
var query = table.CreateQuery().WithParameter("userId", userId);
await table.PullItemsAsync(query, new PullOptions { QueryId = $"q-{tableName}-{userId}" });
```

What will happen is that the GET operation will have parameter userId attached to it.  You can now access the userId in the access control provider, and limit (with GetDataView) the records that are returned any which way you want.

**Reply by jlantech** [reply]
Hello Adrian.

 Again, thank you for you response, I implement the access control provider, but its appear that I'm using the wrong version on my project?, this is a Xamarin.Forms application and the shared project are NetStandard 2.0. 
  this is the call from the client.


                   ```
                     await deviceConfigTable.PullItemsAsync(query:$"$filter=USER_ID eq '{userId}'");
                    await erpcnfTable.PullItemsAsync(query: $"$filter=USER_ID eq '{userId}'");
                    await docnumTable.PullItemsAsync(query: $"$filter=USER_ID eq '{userId}'");   
                 
                    var query = customerTable.CreateQuery().WithParameter("userId", userId);
                    await customerTable.PullItemsAsync(query, new PullOptions {  QueryId = $"q-{customerTable.TableName}-{userId}" });
```

 The customerTable, generate the query , but on the server the `userId` is not visible on the `HttpContextAccessor`, and the client end with an error in thet customerTable.PullItemSync process.

**Reply by adrianhall** [reply]
The HttpContextAccessor is used to access the HttpContext.  Once you have the HttpContext, you have access to the context.Request.  See https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-7.0

Once you have the request, you can access the Query property (which is a QueryCollection) and should be able to access the userId there.  You should, at a minimum, see the other parameters being sent (skip/top/filter/etc.)

And on another note, you should NOT be sending the userId like this.  It leaves your service open to abuse.  Use authentication and grab the userId from the authorization token instead.  This ensures that the user can only see the data that they are allowed to see.
